File: web/src/crypto/envelope.ts
import { concat } from "./bytes";
import { aeadDecrypt, aeadEncrypt, randomBytes } from "./provider";
import { CryptoError } from "./sodium";

/** spec 02 §2 — version ‖ alg ‖ nonce(24) ‖ ct+tag; mirrors core Envelope.kt. */
export const ENVELOPE_VERSION = 0x01;
export const ENVELOPE_ALG_XCHACHA20POLY1305_IETF = 0x01;
export const NONCE_BYTES = 24;
export const TAG_BYTES = 16;
const HEADER_BYTES = 2 + NONCE_BYTES;
export const MIN_BYTES = HEADER_BYTES + TAG_BYTES;

export function sealWithNonce(key: Uint8Array, nonce: Uint8Array, plaintext: Uint8Array, ad: Uint8Array): Uint8Array {
  if (nonce.length !== NONCE_BYTES) throw new CryptoError("nonce must be 24 bytes");
  const ct = aeadEncrypt(key, nonce, plaintext, ad);
  return concat(Uint8Array.of(ENVELOPE_VERSION, ENVELOPE_ALG_XCHACHA20POLY1305_IETF), nonce, ct);
}

export function seal(key: Uint8Array, plaintext: Uint8Array, ad: Uint8Array): Uint8Array {
  return sealWithNonce(key, randomBytes(NONCE_BYTES), plaintext, ad);
}

/**
 * The three STRUCTURAL refusals {@link open} makes before any key is used — length, version, alg —
 * as a reason string (null = the header is well-formed). TWIN of core Envelope.structuralRefusal.
 *
 * Single-sourced here (open() calls it) so a caller that must tell "this blob is damaged/foreign"
 * apart from "this key is wrong" never re-implements — and drifts from — the checks below. WHY the
 * distinction is security-relevant (audit H67): only the AEAD tag failure at the end of open() is
 * genuinely ambiguous — it is exactly what a wrong password produces. These three are decided from
 * the blob's PUBLIC header, before the key matters, so folding them into a "wrong password" verdict
 * tells a member with a corrupt or newer-version account row to reset a password that was never
 * wrong. Account-key callers (Account.unlock's wrappedUvk) consult this first; every other caller
 * keeps the plain CryptoError fail-closed behaviour, because for an item blob "damaged" and "not
 * for this key" are equally unactionable to the user.
 */
export function structuralRefusal(envelope: Uint8Array): string | null {
  if (envelope.length < MIN_BYTES) return "envelope too short";
  if (envelope[0] !== ENVELOPE_VERSION) return `unknown envelope version ${envelope[0]}`;
  if (envelope[1] !== ENVELOPE_ALG_XCHACHA20POLY1305_IETF) return `unknown envelope alg ${envelope[1]}`;
  return null;
}

export function open(key: Uint8Array, envelope: Uint8Array, ad: Uint8Array): Uint8Array {
  const refusal = structuralRefusal(envelope);
  if (refusal !== null) throw new CryptoError(refusal);
  const nonce = envelope.subarray(2, HEADER_BYTES);
  const ct = envelope.subarray(HEADER_BYTES);
  return aeadDecrypt(key, nonce, ct, ad);
}

File: core/src/commonMain/kotlin/io/silencelen/andvari/core/crypto/Envelope.kt
package io.silencelen.andvari.core.crypto

/**
 * spec 02 §2 — the AEAD envelope: version ‖ alg ‖ nonce(24) ‖ ct+tag(≥16),
 * transported as unpadded base64url.
 */
object Envelope {
    const val VERSION: Byte = 0x01
    const val ALG_XCHACHA20POLY1305_IETF: Byte = 0x01
    const val NONCE_BYTES = 24
    const val TAG_BYTES = 16
    private const val HEADER_BYTES = 2 + NONCE_BYTES
    const val MIN_BYTES = HEADER_BYTES + TAG_BYTES

    /** Encrypt with an explicit nonce — used by vector generation; production callers use [seal]. */
    fun sealWithNonce(
        crypto: CryptoProvider,
        key: ByteArray,
        nonce: ByteArray,
        plaintext: ByteArray,
        ad: ByteArray,
    ): ByteArray {
        require(nonce.size == NONCE_BYTES) { "nonce must be $NONCE_BYTES bytes" }
        val ct = crypto.aeadEncrypt(key, nonce, plaintext, ad)
        return byteArrayOf(VERSION, ALG_XCHACHA20POLY1305_IETF) + nonce + ct
    }

    fun seal(crypto: CryptoProvider, key: ByteArray, plaintext: ByteArray, ad: ByteArray): ByteArray =
        sealWithNonce(crypto, key, crypto.randomBytes(NONCE_BYTES), plaintext, ad)

    /**
     * The three STRUCTURAL refusals [open] makes before any key is used — length, version, alg —
     * as a reason string (null = the header is well-formed), single-sourced here so a caller that
     * must tell "this blob is damaged/foreign" apart from "this key is wrong" never re-implements
     * (and drifts from) the checks below.
     *
     * WHY the distinction is security-relevant (audit H67): only the AEAD tag failure at the end of
     * [open] is genuinely ambiguous — it is what a wrong password looks like. These three depend
     * solely on the blob's PUBLIC header and are decided before the key matters, so folding them
     * into a "wrong password" verdict tells a member with a corrupt or newer-version account row to
     * go and reset a password that was never wrong. Callers holding an ACCOUNT-KEY blob
     * (Account.unlock's wrappedUvk) consult this first; every other caller keeps the plain
     * [CryptoException] fail-closed behaviour, because for an item blob "damaged" and "not for this
     * key" are equally unactionable to the user.
     */
    fun structuralRefusal(envelope: ByteArray): String? = when {
        envelope.size < MIN_BYTES -> "envelope too short"
        envelope[0] != VERSION -> "unknown envelope version ${envelope[0]}"
        envelope[1] != ALG_XCHACHA20POLY1305_IETF -> "unknown envelope alg ${envelope[1]}"
        else -> null
    }

    fun open(crypto: CryptoProvider, key: ByteArray, envelope: ByteArray, ad: ByteArray): ByteArray {
        structuralRefusal(envelope)?.let { throw CryptoException(it) }
        val nonce = envelope.copyOfRange(2, HEADER_BYTES)
        val ct = envelope.copyOfRange(HEADER_BYTES, envelope.size)
        return crypto.aeadDecrypt(key, nonce, ct, ad)
    }

    fun sealB64(crypto: CryptoProvider, key: ByteArray, plaintext: ByteArray, ad: ByteArray): String =
        Bytes.toB64(seal(crypto, key, plaintext, ad))

    fun openB64(crypto: CryptoProvider, key: ByteArray, envelopeB64: String, ad: ByteArray): ByteArray =
        open(crypto, key, Bytes.fromB64(envelopeB64), ad)
}
